File: project-signer-runnable/src/main/kotlin/org/jesperancinha/parser/projectsigner/ProjectSignerStart.java
package org.jesperancinha.parser.projectsigner;

import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.jesperancinha.parser.projectsigner.configuration.ProjectSignerOptions;
import org.jesperancinha.parser.projectsigner.service.FileWriterService;
import org.jesperancinha.parser.projectsigner.service.FinderService;
import org.jesperancinha.parser.projectsigner.service.OptionsService;
import org.jesperancinha.parser.projectsigner.service.ReadmeService;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.core.env.Environment;

import java.util.Arrays;

@Slf4j
@SpringBootApplication
public class ProjectSignerStart implements ApplicationRunner {

    private final Environment environment;
    private final FinderService finderService;
    private final FileWriterService fileWriterService;
    private final OptionsService optionsService;
    private final ReadmeService readmeService;

    public static void main(String[] args) {
        SpringApplication.run(ProjectSignerStart.class, args);
    }

    public ProjectSignerStart(final FinderService finderService,
                              final FileWriterService fileWriterService,
                              final OptionsService optionsService,
                              final ReadmeService readmeService, Environment environment) {
        this.finderService = finderService;
        this.fileWriterService = fileWriterService;
        this.optionsService = optionsService;
        this.readmeService = readmeService;
        this.environment = environment;
    }

    @Override
    public void run(ApplicationArguments args) throws Exception {
        if (environment.getActiveProfiles().length > 0) {
            final ProjectSignerOptions projectSignerOptions = optionsService.processOptions(args.getSourceArgs());
            finderService.iterateThroughFilesAndFolders(projectSignerOptions.getRootDirectory());
            val profiles = Arrays.asList(environment.getActiveProfiles());
            if (profiles.contains("prod") || profiles.contains("dev")) {
                fileWriterService.exportReportFiles(
                        optionsService.getProjectSignerOptions().getReportLocation(), readmeService.getAllProjectData());
            }
        }
    }
}

File: project-signer-runnable/src/main/kotlin/org/jesperancinha/parser/projectsigner/ProjectSignerStart.kt
package org.jesperancinha.parser.projectsigner

import org.jesperancinha.parser.projectsigner.service.FileWriterService
import org.jesperancinha.parser.projectsigner.service.FinderService
import org.jesperancinha.parser.projectsigner.service.OptionsService
import org.jesperancinha.parser.projectsigner.service.ReadmeService
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.core.env.Environment
import java.nio.file.Path
import kotlin.Throws

@SpringBootApplication
open class ProjectSignerStart(
    private val finderService: FinderService,
    private val fileWriterService: FileWriterService,
    private val optionsService: OptionsService,
    private val readmeService: ReadmeService,
    private val environment: Environment
) : ApplicationRunner {
    @Throws(Exception::class)
    override fun run(args: ApplicationArguments) {
        System.setProperty("file.encoding","UTF-8")
        if (environment.activeProfiles.isNotEmpty()) {
            val projectSignerOptions = optionsService.processOptions(args.sourceArgs)
            finderService.iterateThroughFilesAndFolders(projectSignerOptions.rootDirectory ?: throw RuntimeException("Root directory needs to be configured!"))
            val profiles = listOf(*environment.activeProfiles)
            if (profiles.contains("prod") || profiles.contains("dev")) {
                optionsService.projectSignerOptions?.reportLocation?.let { Path.of(it) ?: throw RuntimeException("Report location needs to be configured") }
                    ?.let {
                        fileWriterService.exportReportFiles(
                            it, readmeService.allProjectData
                        )
                    }
            }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            SpringApplication.run(ProjectSignerStart::class.java, *args)
        }
    }
}